src/main/java/org/opensearch/knn/index/memory/NativeMemoryAllocation.java

```diff
@@ -328,8 +328,8 @@ public TrainingDataAllocation(ExecutorService executor, long memoryAddress, int
         @Override
         public void close() {
             executor.execute(() -> {
+                writeLock();
                 try {
-                    writeLock();
```

Comment:
If an exception was thrown there, it will called unmatched unlock call in finally block.
I know the current implementation is not throwing any exceptions. Here I was trying to make the code robust no matter how other parts changed.